ec/ecresource/deploymentresource/planmodifiers/topology_dedicated_masters.go

```diff
@@ -0,0 +1,189 @@
+			tflog.Debug(ctx, "UpdateDedicatedMasterTier: Could not enable master tier, as it has no instance-config.")
+			return
+		}
+
+		if instanceConfiguration.DiscreteSizes == nil {
+			return
+		}
+
+		// Set zones
+		zones := instanceConfiguration.MaxZones
+		if zones == 0 {
+			return
+		}
+		resp.Plan.SetAttribute(ctx,
+			path.Root("elasticsearch").AtName("master").AtName("zone_count"),
+			zones,
+		)
+
+		// Set Size
+		defaultSize := util.MemoryToState(instanceConfiguration.DiscreteSizes.DefaultSize)
+		resp.Plan.SetAttribute(ctx,
+			path.Root("elasticsearch").AtName("master").AtName("size"),
+			defaultSize,
+		)
+	}
+}
+
+func getInstanceConfiguration(template models.DeploymentTemplateInfoV2) *models.InstanceConfigurationInfo {
+	// Find master tier
+	if template.DeploymentTemplate == nil ||
+		template.DeploymentTemplate.Resources == nil ||
+		len(template.DeploymentTemplate.Resources.Elasticsearch) == 0 ||
+		template.DeploymentTemplate.Resources.Elasticsearch[0].Plan == nil {
+		return nil
+	}
+	var masterTier *models.ElasticsearchClusterTopologyElement
+	for _, topology := range template.DeploymentTemplate.Resources.Elasticsearch[0].Plan.ClusterTopology {
+		if topology.ID == "master" {
+			masterTier = topology
+		}
+	}
+	if masterTier == nil {
+		return nil
+	}
+
+	// Find IC for master tier
+	for _, ic := range template.InstanceConfigurations {
+		if ic.ID == masterTier.InstanceConfigurationID {
+			return ic
+		}
+	}
+
+	return nil
+}
+
+func getDedicatedMastersThreshold(template models.DeploymentTemplateInfoV2) int64 {
+	dedicatedMastersThreshold := 0
+	if template.DeploymentTemplate != nil && template.DeploymentTemplate.Resources != nil {
+		for _, e := range template.DeploymentTemplate.Resources.Elasticsearch {
+			if e.Settings == nil {
+				continue
+			}
+			dedicatedMastersThreshold = int(e.Settings.DedicatedMastersThreshold)
+			break
+		}
+	}
+	return int64(dedicatedMastersThreshold)
+}
+
+func countNodesInCluster(
+	ctx context.Context,
+	planElasticsearch es.ElasticsearchTF,
+) int64 {
+	nodesInDeployment := int64(0)
+	tiers := []types.Object{
+		planElasticsearch.HotContentTier,
+		planElasticsearch.CoordinatingTier,
+		planElasticsearch.WarmTier,
+		planElasticsearch.ColdTier,
+		planElasticsearch.FrozenTier,
+	}
+	for _, tier := range tiers {
+		if tier.IsNull() || tier.IsUnknown() {
+			continue
+		}
+
+		var topology es.ElasticsearchTopologyTF
+		diags := tier.As(ctx, &topology, objectAsOptions)
+		if diags.HasError() {
+			tflog.Debug(ctx, "countNodesInCluster: Failed to read topology.", map[string]interface{}{"error": diags.Errors()})
+			continue
+		}
+
+		size, err := deploymentsize.ParseGb(topology.Size.ValueString())
+		if err != nil {
+			tflog.Debug(ctx, "countNodesInCluster: Failed to parse topology size.", map[string]interface{}{"error": err})
+			continue
+		}
+
+		zoneCount := topology.ZoneCount.ValueInt64()
+		if size > 0 && zoneCount > 0 {
+			nodesInDeployment += zoneCount
+		}
```

Comment:
How does this handle the case when `size` is greater than the maximum node size? That is there are multiple nodes for a given tier within a zone?

Reply:
That's a very good question. I have to double check that since it looks like it wouldn't cover that case.

Reply:
I think the logic should be (from looking at how the UI does it): 
- Use `node_count_per_zone` if available
- Otherwise estimate it with `size / max_size`

Reply:
Actually `node_count_per_zone` (Which is supposed to be combined with `memory_per_node`) is unsupported in the provider: It will just ignore those values if set and use the template instead.

So my solution will be to calculate the max from the IC and use that to calculate the nodes-per-zone